### DISCUSSION TITLE
Infra CLI

### DISCUSSION THREAD

**Discussion by codyebberson** [medplum/medplum#1839] [Ideas]
# Infra CLI

This discussion is about command line tools to simplify the process of managing a self-hosted Medplum cluster on AWS.

## Goals

* Devops tools to easily create and maintain Medplum deployments
* Best practices for managing configuration state
* Best practices for upgrading to the latest version of Medplum
* Best practices for upgrading cluster configuration

## Current state

* Good:
  * Medplum provides the CDK "init" tool which can generate a JSON file with CDK settings
  * Deploy new version of server with single command
  * Everything is ***possible*** today via command line
  * Medplum hosted service has run with true CI/CD for almost 2 years without issue
* ~~Bad~~ Room for improvement:
  * Configuration state lives in multiple different places
    * CDK config lives in JSON config file
    * Server config lives in AWS Parameter Store
    * App config lives in `.env` file
  * Deploying new version is not straightforward
    * CDK updates require CDK commands
      * `npx cdk -c config=medplum.example.config.json`
    * Server updates require AWS CLI commands with magical values
      * `aws ecs update-cluster --region xyz --cluster xyz --service xyz --force-new-deployment`
    * App updates require cloning repo and running full build (!!!)
      * `git pull origin master`
      * `npm run build -- --force`
      * `APP_BUCKET=app.medplum.example.com ./scripts/deploy-app.sh` 

## Proposal

Add collection of commands to the Medplum CLI specifically for managing AWS deployments.

`medplum aws init`

1. Move the current CDK `init` tool into the CLI project
2. Writes local config files for all services
  a. Single "config" file for cdk, server, and app?
  b. Multiple config files, one each for cdk, server, and app?
3. Add options to actually run the CDK commands, or to print CDK commands to manually run

`medplum aws list`

1. Query AWS CloudFormation for Medplum deployments
2. Map the deployment "name" to the AWS values

`medplum aws upgrade [name] [latest | version]`

1. Updates the Medplum Deployment to the version (either specific version tag or "latest")
2. Runs CDK diff to check for infra changes
3. Updates ECS cluster
  a. Maps "name" to ECS "Cluster" and "Service"
  b. If setting a specific version, updates the AWS ECS Task Definition
  c. Runs `aws ecs update-service`
4. Deploys `app` to the S3 bucket
  a. Downloads the pre-built app from npmjs.com (?)
  b. Dynamically rewrites config settings with cluster specific values

`medplum aws destroy [name]`

Permanently deletes the AWS CloudFormation stack

## Questions

1. We need to settle on name for a Medplum "deployment".
  a. "Medplum Deployment"
  b. "Medplum Service"
  c. "Medplum Cluster"
2. Should the Medplum commands invoke the CDK directly?  Or should they sit "beside" the CDK?
  a. We could do both, and give users the option to run automatically or to print the commands to run manually.
3. How can we determine which CloudFormation stacks are Medplum stacks?
  a. We could try to do it by name
  b. We may want to add AWS "tags" of some sort to make it concrete

**Comment by rahul1** [comment]
> We need to settle on name for a Medplum "deployment".
I would choose "cluster" or "instance". "service" is pretty overloaded, especially with users who have microservice architectures.
